File: courses/focs-fa20/homeworks/homework5.ml
(* 

HOMEWORK 5

Due: Thu Nov 5, 2020 (23h59)

Name: 

Email:

Remarks, if any:

*)


(*
 *
 * Please fill in this file with your solutions and submit it
 *
 * The functions below are stubs that you should replace with your
 * own implementation.
 *
 * PLEASE DO NOT CHANGE THE TYPES IN THE STUBS BELOW.
 * Doing so risks making it impossible for me to test your code.
 *
 * Always make sure you can #use this file in a FRESH OCaml shell
 * before submitting it. It has to load without any errors.
 *
 *)


(*
 *  The type for instructions - each row is a constructor representing an instruction
 *
 *) 

type instruction =
  | INC of string
  | DEC of string * string
  | JUMP of string
  | TRUE
  | FALSE
  | PRINT of string
  | LABEL of string
  | REGISTER of string * int
  | EQUAL of string * string

let pp instructions =
  let instruction_to_string i instr =
    match instr with
    | INC n -> Printf.sprintf "%04d INC %s\n" i n
    | DEC (n, addr) -> Printf.sprintf "%04d DEC (%s, %s)\n" i n addr
    | JUMP addr -> Printf.sprintf "%04d JUMP %s\n" i addr
    | TRUE -> Printf.sprintf "%04d TRUE\n" i
    | FALSE -> Printf.sprintf "%04d FALSE\n" i
    | PRINT n -> Printf.sprintf "%04d PRINT %s\n" i n
    | LABEL addr -> Printf.sprintf "%04d LABEL %s\n" i addr
    | REGISTER (r, n) -> Printf.sprintf "%04d REGISTER (%s, %d)\n" i r n
    | EQUAL (r1, r2) -> Printf.sprintf "%04d EQUAL (%s, %s)\n" i r1 r2  in
  let program_to_string instructions =
    String.concat "" (List.mapi instruction_to_string instructions)  in
  print_string (program_to_string instructions)


(* 
 * Helper functions
 *
 * Create a map from registers name to positions
 * Create a map from label names to instruction indices
 * Lookup a name in a map
 * Get the largest register used in a program
 *
 *)
  
module SMap = Map.Make(String)
    
let registerMap (instrs: instruction list): int SMap.t =
  let rec loop instrs map =
    match instrs with
    | [] -> map
    | (REGISTER (name, r))::instrs' -> loop instrs' (SMap.add name r map)
    | _::instrs' -> loop instrs' map  in
  loop instrs SMap.empty
    
let labelMap (instrs: instruction list): int SMap.t = 
  let rec loop instrs addr map =
    match instrs with
    | [] -> map
    | (LABEL n)::instrs' -> loop instrs' (addr + 1) (SMap.add n addr map)
    | _::instrs' -> loop instrs' (addr + 1) map  in
  loop instrs 0 SMap.empty

let lookup sym map =
  try SMap.find sym map with
  | Not_found -> let _ = Printf.printf "Cannot find symbol %s\n" sym in
                 raise Not_found

let max_register (p : instruction list): int =
  let rec loop instrs curr_max =
    match instrs with
    | [] -> curr_max
    | (REGISTER (_, cntr))::instrs' -> loop instrs' (max curr_max cntr)
    | _::instrs' -> loop instrs' curr_max in
  loop p 0

  
(*
 * Run a CPU program directly
 *
 * Inputs are: program and list of initial values for the first few registers
 *
 *)
  
let run (p: instruction list) (nums: int list): bool =
  let _ = print_string "----------------------------------------------------------------------\n" in
  let _ = pp p in
  let _ = print_string "----------------------------------------------------------------------\n" in
  let rmap = registerMap p in
  let lmap = labelMap p in
  let p_array = Array.of_list p in
  let registers =  Array.make (max_register p + 1) 0 in
  let _ = List.iteri (fun i v -> Array.set registers i v) nums in
  let rec loop addr =
    let regs = String.concat " " (List.map string_of_int (Array.to_list registers)) in
    let _ = Printf.printf "%04d: %s\n" addr regs in
    match (Array.get p_array addr) with
    | INC (r) -> let r = lookup r rmap in
                 let _ = Array.set registers r ((Array.get registers r) + 1) in
                 loop (addr + 1) 
    | DEC (r, addr') -> let r = lookup r rmap in
                        let addr' = lookup addr' lmap in
                        let v = Array.get registers r in
                        if (v > 0) then
                          let _ = Array.set registers r (v - 1) in
                          loop (addr + 1)
                        else
                          loop addr'
    | JUMP addr' -> let addr' = lookup addr' lmap in
                    loop addr'
    | TRUE -> true
    | FALSE -> false
    | PRINT r -> let r' = lookup r rmap in
                 let _ = Printf.printf "      Register %s = %d\n" r (Array.get registers r') in
                 loop (addr + 1)
    | EQUAL (r1, r2) -> let r1 = lookup r1 rmap in
                        let r2 = lookup r2 rmap in
                        Array.get registers r1 = Array.get registers r2
    | _ -> loop (addr + 1)  in
  loop 0


(*
 *
 * Some very simple programs
 *
 *)
 

let p_reset = [REGISTER ("X", 0);
               LABEL "loop";
               DEC ("X", "done");
               JUMP "loop";
               LABEL "done";
               TRUE]

let p_transfer = [REGISTER ("X", 0);
                  REGISTER ("Y", 1);
                  LABEL "loop";
                  DEC ("X", "done");
                  INC "Y";
                  JUMP "loop";
                  LABEL "done";
                  TRUE]

let p_succ = [REGISTER ("X", 0);
              REGISTER ("Z", 1);
	      INC "X";
	      EQUAL ("X", "Z")]



(* QUESTION 1 *)


let p_copy: instruction list = [REGISTER ("X", 0);
                                REGISTER ("Y", 1)]


let p_swap: instruction list = [REGISTER ("X", 0);
                                REGISTER ("Y", 1)]


let p_plus: instruction list = [REGISTER ("X", 0);
                                REGISTER ("Y", 1);
                                REGISTER ("Z", 2)]


let p_sub: instruction list = [REGISTER ("X", 0);
                               REGISTER ("Y", 1);
                               REGISTER ("Z", 2)]


let p_max: instruction list = [REGISTER ("X", 0);
                               REGISTER ("Y", 1);
                  	     REGISTER ("Z", 2)]
             
           
let p_diff: instruction list = [REGISTER ("X", 0);
                                REGISTER ("Y", 1);
                                REGISTER ("Z", 2)]
               

(* QUESTION 2 *)


let p_times: instruction list = [REGISTER ("X", 0);
                                 REGISTER ("Y", 1);
                                 REGISTER ("Z", 2)]


let p_square: instruction list = [REGISTER ("X", 0);
                                  REGISTER ("Z", 1)]


let p_square_x_2: instruction list = [REGISTER ("X", 0);
                                      REGISTER ("Z", 1)]
            

(* QUESTION 3 *)

let p_half: instruction list = [REGISTER ("X", 0);
                                REGISTER ("Z", 1)]

File: courses/focs-fa24/homeworks/5/homework5.py
# --------------------------------------------------
#
# HOMEWORK 5
#
# Due: Tue, Nov 19 2024 (23h59)
#
# Name:
#
# Email:
#
# Remarks, if any:
#
#
#
# --------------------------------------------------
#
# Please fill in this file with your solutions and submit it
#
# The functions below are stubs that you should replace with your
# own implementation.
#
# PLEASE DO NOT CHANGE THE SIGNATURE IN THE STUBS BELOW.
# Doing so risks making it impossible for me to test your code.
#
# --------------------------------------------------



class RM:

    def __init__(self):
        self._silent = False
        self._fresh = 100

    def silent(self, flag):
        self._silent = flag

    def expand_memory(self, register, memory):
        if register >= len(memory):
            memory.extend([0] * (1 + register - len(memory)))

    def show_step(self, pc, memory):
        content = "  ".join(str(c) for c in memory)
        if pc is None:
            print(f"----  {content}")
        else:
            print(f"{pc:04d}  {content}")

    def show_program(self, program):
        pc = 0
        for instruction in program:
            content = " ".join(str(c) for c in instruction)
            print(f"{pc:04d}  {content}")
            if instruction[0] not in ["label", "register", "temporary"]:
                pc += 1

    def get_register(self, memory, register):
        if register >= len(memory):
            memory.extend([0] * (1 + register - len(memory)))
        return memory[register]

    def max_register(self, program):
        def get_reg(instruction):
            opcode = instruction[0]
            if opcode == "inc" and isinstance(instruction[1], int):
                return instruction[1]
            elif opcode == "dec" and isinstance(instruction[1], int):
                return instruction[1]
            elif opcode == "stop" and isinstance(instruction[1], int):
                return instruction[1]
            elif opcode == "register" and isinstance(instruction[2], int):
                return instruction[2]
            else:
                return 0
        mx = 0
        for p in program:
            mx = max(mx, get_reg(p))
        return mx

    def run(self, program, init):
        program = self.resolve_registers(program)
        program = self.resolve_labels(program)
        pc = 0
        memory = init
        stop = None
        while stop is None:
            if pc < 0 or pc >= len(program):
                raise Exception(f"@{pc}: attempting to access instruction outside the program (= segfault)")
            if not self._silent:
                self.show_step(pc, memory)
            instruction = program[pc]
            try:
                opcode = instruction[0].lower()
                if opcode == 'inc':
                    v = self.get_register(memory, instruction[1])
                    memory[instruction[1]] = v + 1
                    pc += 1
                elif opcode == 'dec':
                    self.expand_memory(instruction[1], memory)
                    v = self.get_register(memory, instruction[1])
                    if v == 0:
                        pc = instruction[2]
                    else:
                        memory[instruction[1]] = v - 1
                        pc += 1
                elif opcode == 'jump':
                    pc = instruction[1]
                elif opcode == 'stop':
                    stop = self.get_register(memory, instruction[1])
                else:
                    raise Exception
            except:
                raise Exception(f"At {pc:04d}: bad instruction {instruction}")
        self.show_step(None, memory)
        return stop

    def resolve_registers(self, program):
        # Complete this method for Question 2(A).
        # Currently it does nothing.
        return program

    def resolve_labels(self, program):
        # Complete this method for Question 2(B).
        # Currently, it does nothing.
        return program




# Some sample programs.

P_RESET = [
    # Reset register 0.
    ("dec", 0, 2),
    ("jump", 0),

    ("stop", 0)
]

P_TRANSFER = [
    # Transfer register 0 to register 1.
    ("dec", 0, 3),
    ("inc", 1),
    ("jump", 0),

    ("stop", 0)
]

P_COPY = [
    # Copy register 0 to register 1 (preserving 0).
    ("dec", 0, 4),
    ("inc", 1),
    ("inc", 2),
    ("jump", 0),

    ("dec", 2, 7),
    ("inc", 0),
    ("jump", 4),

    ("stop", 1)
]

P_PLUS1 = [
    # Z := X + 1
    ("dec", 0, 3),
    ("inc", 2),
    ("jump", 0),

    ("dec", 2, 7),
    ("inc", 0),
    ("inc", 1),
    ("jump", 3),

    ("inc", 1),
    ("stop", 1)
]


# Question 1
#
# Replace these placeholder programs by your own.

P_PLUS = [
    ("stop", 0)
]

P_MINUS = [
    ("stop", 0)
]

P_MAX = [
    ("stop", 0)
]


# Question 2
#
# Replace these placeholder programs by your own.

P_DIFF = [
    ("register", "X", 0),
    ("register", "Y", 1),
    ("register", "Z", 2),
    ("stop", "Z")
]
